base/channels.jl

```diff
@@ -457,3 +457,21 @@ function iterate(c::Channel, state=nothing)
 end
 
 IteratorSize(::Type{<:Channel}) = SizeUnknown()
+
+"""
+    Threads.foreach(f, channel::Channel; ntasks=Threads.nthreads())
+
+Similar to `foreach(f, channel)`, but iteration over `channel` and calls to
+`f` are split across `ntasks` Tasks spawned by `Threads.@spawn`. This function
+will wait for all internally spawned tasks to complete before returning.
+"""
+function Threads.foreach(f, channel::Channel; ntasks=Threads.nthreads())
+    @sync for _ in 1:ntasks
+        Threads.@spawn begin
+            for item in channel
+                wait(Threads.@spawn f(item))
```

Comment:
I just realized there is a memory leak and thread-unsafe `push!` (I think). With this code as-is, `Threads.@spawn f(item)` keeps `push!`ing the tasks to the task list of the outer `@sync`. I think we need to declare local lexical scope like this:

```suggestion
                @sync Threads.@spawn f(item)
```

Reply:
ref https://github.com/JuliaLang/julia/issues/34666

Reply:
Ah, really good catch. Just got rid of the `@sync` and manually called `Base.sync_end` on the tasks for now.